boa_engine/src/module/mod.rs

```diff
@@ -0,0 +1,701 @@
+    }
+
+    /// Abstract operation [`GetModuleNamespace ( module )`][spec].
+    ///
+    /// Gets the [**Module Namespace Object**][ns] that represents this module's exports.
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-getmodulenamespace
+    /// [ns]: https://tc39.es/ecma262/#sec-module-namespace-exotic-objects
+    pub fn namespace(&self, context: &mut Context<'_>) -> JsObject {
+        self.inner
+            .namespace
+            .borrow_mut()
+            .get_or_insert_with(|| {
+                let exported_names = self.get_exported_names(&mut Vec::default());
+
+                let unambiguous_names = exported_names
+                    .into_iter()
+                    .filter_map(|name| {
+                        self.resolve_export(name, &mut HashSet::default())
+                            .ok()
+                            .map(|_| name)
+                    })
+                    .collect();
+
+                ModuleNamespace::create(self.clone(), unambiguous_names, context)
+            })
+            .clone()
+    }
+}
+
+impl PartialEq for Module {
+    fn eq(&self, other: &Self) -> bool {
+        std::ptr::eq(self.inner.as_ref(), other.inner.as_ref())
+    }
+}
+
+impl Eq for Module {}
+
+impl Hash for Module {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        std::ptr::hash(self.inner.as_ref(), state);
+    }
+}
+
+/// Module namespace exotic object.
+///
+/// Exposes the bindings exported by a [`Module`] to be accessed from ECMAScript code.
+#[derive(Debug, Trace, Finalize)]
+pub struct ModuleNamespace {
+    module: Module,
+    #[unsafe_ignore_trace]
+    exports: IndexMap<JsString, Sym, BuildHasherDefault<FxHasher>>,
+}
+
+impl ModuleNamespace {
+    /// Abstract operation [`ModuleNamespaceCreate ( module, exports )`][spec].
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-modulenamespacecreate
+    pub(crate) fn create(module: Module, names: Vec<Sym>, context: &mut Context<'_>) -> JsObject {
+        let mut exports = names
+            .into_iter()
+            .map(|sym| {
+                (
+                    context
+                        .interner()
+                        .resolve_expect(sym)
+                        .into_common::<JsString>(false),
+                    sym,
+                )
+            })
+            .collect::<IndexMap<_, _, _>>();
+        exports.sort_keys();
+
+        let namespace = JsObject::from_proto_and_data_with_shared_shape(
+            context.root_shape(),
+            None,
+            ObjectData::module_namespace(ModuleNamespace { module, exports }),
+        );
+
+        namespace.borrow_mut().properties_mut().insert(
+            &PropertyKey::Symbol(JsSymbol::to_string_tag()),
+            PropertyDescriptor::builder()
+                .value(js_string!("Module"))
+                .writable(false)
+                .enumerable(false)
+                .configurable(false)
+                .build(),
+        );
```

Comment:
We could have an `ObjectTemplate` so we direct initialize it and not have to add properties after creation.